db/migrate/20190222000434_create_event_logs.rb

```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class CreateEventLogs < ActiveRecord::Migration
+  def change
+    create_table :event_logs do |t|
+      t.uuid :request_id, null: false, unique: true
+      t.string :type
+      t.string :ip_address
+      t.string :state
+      t.string :description
+      t.jsonb  :data
+
+      # Optional Foreign Key Ids
```

Comment:
Could the solution be implemented without single table inheritance? I've always regretted it later when I've used it. One reason is you often end up with optional/nullable columns (now and especially in the future if a sub-class needs a new field). Gitlab's advice on STI in their development guide [is avoid at all costs](https://docs.gitlab.com/ee/development/single_table_inheritance.html) :)

Reply:
I'm sure there is a way to avoid it. And I think for now I will maybe try to come up with a solution that might solve this problem with just a "SessionsLog" table since our case is a little bit unique in having to link a future event with a previous event (saml callback with init). 

I'll definitely take a look at the links you provided up above and update the RFC to reflect some of those. I think I do like that approach better. Personally I've always loved STI, it's object orientated programming at its finest -- the null columns, particularly adding columns and indexes on columns when you have millions of records though is a beast so I generally prefer if an STI structure is well defined from the onset... future proofed. But yea, there are limitations and grievances to be had if its not done right.